File: DSA/Pragya_Agarwal/Day4_Question1.py
string = "ababa"
length = 3
n = len(string)
palindromes = set()

for i in range(n - length + 1):
    substring = string[i:i+length]
    if substring == substring[::-1]:
        palindromes.add(substring)
        
result = len(palindromes)
print("Number of distinct palindromic substrings of length",length,":",result)

File: DSA/Manav-Rajput/Day4_question_1.java
import java.util.HashSet;

public class PalindromicSubstrings {
    public static int countPalindromicSubstrings(String str, int length) {
        HashSet<String> palindromes = new HashSet<>();

        for (int i = 0; i <= str.length() - length; i++) {
            String substr = str.substring(i, i + length);

            if (isPalindrome(substr)) {
                palindromes.add(substr);
            }
        }

        return palindromes.size();
    }

    public static boolean isPalindrome(String str) {
        int start = 0;
        int end = str.length() - 1;

        while (start < end) {
            if (str.charAt(start) != str.charAt(end)) {
                return false;
            }

            start++;
            end--;
        }

        return true;
    }

    public static void main(String[] args) {
        String str = "ababa";
        int length = 3;

        System.out.println("Number of distinct palindromic substrings: " + countPalindromicSubstrings(str, length));
    }
}
